example/amd_smi_list.cc

```diff
@@ -0,0 +1,169 @@
+
+    std::cout << "Total Socket: " << socket_count << std::endl;
+
+    // For each socket, get cpus
+    for (uint32_t i = 0; i < socket_count; i++) {
+
+        // Get the device count available for the socket.
+        uint32_t device_count = 0;
+        uint32_t tmp_device_count = 0;
+        bool isGPU, isNIC, isSWITCH;
+        ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_AMD_GPU, nullptr, &tmp_device_count);
+        CHK_AMDSMI_RET(ret)
+
+        // Allocate the memory for the device handlers on the socket
+        
+        if (tmp_device_count == 0) { //it could be BRCM NIC device
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_NIC, nullptr, &tmp_device_count);
+          device_count = tmp_device_count;
+          isNIC = true;
+          CHK_AMDSMI_RET(ret)
+        } else {
+          device_count = tmp_device_count;
+          isGPU = true;
+        }
+
+        if (tmp_device_count == 0) {  // it could be BRCM SWITCH device
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_SWITCH, nullptr,
+                                                     &tmp_device_count);
+          device_count = tmp_device_count;
+          isNIC = false;
+          isGPU = false;
+          isSWITCH = true;
+          CHK_AMDSMI_RET(ret)
+        }
+
+        // Allocate the memory for the device handlers on the socket
+        std::vector<amdsmi_processor_handle> processor_handles(device_count);
+
+        if (isNIC) {
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_NIC, &processor_handles[0], &device_count);
+          CHK_AMDSMI_RET(ret)
+        } else if (isGPU) {
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_AMD_GPU, &processor_handles[0], &device_count);
+          CHK_AMDSMI_RET(ret)
+        } else if (isSWITCH) {
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_SWITCH, &processor_handles[0], &device_count);
+          CHK_AMDSMI_RET(ret)
+        } else {
+          printf("\n Unknown device as part of discovered list. Exiting! \n");
+          ret = amdsmi_shut_down();
+          CHK_AMDSMI_RET(ret)
+          return 0;
+        }
+ 
+        for (uint32_t j = 0; j < device_count; j++) {
+
+            processor_type_t processor_type = {};
+            ret = amdsmi_get_processor_type(processor_handles[j], &processor_type);
+            CHK_AMDSMI_RET(ret)
+
+            amdsmi_bdf_t bdf = {};
+
+            if (isNIC) {
+                ret = amdsmi_get_nic_device_bdf(processor_handles[j], &bdf);
+            }
+            else if (isGPU) {
+                ret = amdsmi_get_gpu_device_bdf(processor_handles[j], &bdf);
+            } 
+            else if (isSWITCH) {
+              ret = amdsmi_get_switch_device_bdf(processor_handles[j], &bdf);
+            }
+            std::string pType = "";
```

Comment:
It is a good practice to `#include` the header where it is directly used. So, `#include <string>`.